### DISCUSSION TITLE
Question about a use case for using NRules

### DISCUSSION THREAD

**Discussion by vks2** [NRules/NRules#355] [Q&A] [answered]
dear all. thanks much for that library and well documented examples. however, is there any large open-source proj that implement Nrules? i haven't found any in any field. i'm wondering if there some time saver if i write some custom and rather granular logic of credit decision app from scratch or using rules, that seems to me an overhead if there's no exact need to implement Rete matching. 
to be more precise, why do i need go via
```

public class CreditDecisionRule : Rule
{
    public override void Define()
    {
        Company company = default;
        FinancialData financialData = default;
        CreditHistory creditHistory = default;

        When()
            .Match<Company>(() => company)
            .Match<FinancialData>(() => financialData, f => f.CompanyId == company.Id)
            .Match<CreditHistory>(() => creditHistory, ch => ch.CompanyId == company.Id);

        Then()
            .Do(ctx => CheckFinancialHealth(company, financialData))
            .Do(ctx => CheckCreditHistory(company, creditHistory))
            .Do(ctx => CheckRiskFactors(company));
    }
} 
```
instead of 
 ```
public void MakeCreditDecision(Company company, FinancialData financialData, CreditHistory creditHistory)
    {
        CheckFinancialHealth(company, financialData);
        CheckCreditHistory(company, creditHistory);
        CheckRiskFactors(company);
    }

    private void CheckFinancialHealth(Company company, FinancialData financialData)
    {
        // Logic to analyze financial health and make credit decisions
    }

    private void CheckCreditHistory(Company company, CreditHistory creditHistory)
    {
        // Logic to review credit history and make credit decisions
    }

    private void CheckRiskFactors(Company company)
    {
        // Logic to assess risk factors and make credit decisions
    }
```

**Comment by snikolayev** [comment, marked as answer]
@vks2 thanks for your interest in NRules. Moving forward please ask questions in the discussion forum instead of creating an issue - trying to reserve issues for bug reports and feature requests.
Answering your question, in my experience, the actual business logic is usually proprietary, so I'm not surprised you don't find a lot of it in open source.
I think there are a few things to think about when deciding on how to factor your business logic, such as separation of concerns, externalization, performance, etc. The biggest one in my opinion, to even be considering something like NRules, is the need to decompose the logic up into smaller pieces that can be reasoned about, developed and tested in isolation. Then you need something, like a rules engine, that can orchestrate those pieces (rules) together. Certainly, there is no point in using NRules with one giant block that has no logic in the conditions and just calls three methods. In this case all the logic is opaque and hidden inside those methods and a simple strategy design pattern will be more than enough.
But if you were to break your FinancialData and CreditHistory into smaller constituents, model your domain in a more expressive way, and write different rules that model different aspects of making a risk assessment (e.g. calculating different risk factors) and then maybe a rule that aggregated the calculated risk factors and made the final decision, then you could  benefit from NRules that has a way of orchestrating these different rules together.
